File: 72-edit-distance/edit-distance.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        if len(word2) > len(word1):
            word1 , word2 = word2, word1
        n = len(word1)
        m = len(word2)
        dp = [[float('inf')] * (n+1) for _ in range(m+1)]
        if(n == 0 or m ==0):
            return max(n,m)
        dp[0][0] = 0
        for i in range(m+1):
            for j in range(n+1):
                if (j < n) and dp[i][j+1] > dp[i][j] + 1:
                    dp[i][j+1] = dp[i][j] + 1
                score = dp[i][j] if ((j < n and i < m) and word1[j] == word2[i]) else dp[i][j] + 1
                if (j < n and i < m) and dp[i+1][j+1] > score:
                    dp[i+1][j + 1] = score
                if i < m and dp[i+1][j] > dp[i][j] + 1:
                    dp[i+1][j] = (dp[i][j] + 1)
        # print(dp)
        return dp[m][n]

File: 72-edit-distance/edit-distance.cpp
class Solution {
public:
    int minDistance(string word1, string word2) {
        int n = word1.size();
        int m = word2.size();
        using u32 = __uint32_t;
        constexpr u32 u32m = __UINT32_MAX__;
        vector<vector<u32>> dp (n+1, vector<u32> (m+1, u32m));
        dp[0][0] = 0;
        for(int i = 1; i < m+1; i++){
            dp[0][i] = 1 + dp[0][i-1];
        }
        for(int i = 1; i < n+1; i++){
            dp[i][0] = 1 + dp[i-1][0];
        }

        for(int i = 1; i < n+1; i++) {
            for(int j = 1; j< m+1; j++) {
                if(word1[i-1] == word2[j-1]){
                    dp[i][j] = min(dp[i][j], dp[i-1][j-1]);
                }
                dp[i][j] = min(dp[i][j], 1 + dp[i-1][j-1]);
                dp[i][j] = min(dp[i][j], 1 + dp[i-1][j]);
                dp[i][j] = min(dp[i][j], 1 + dp[i][j-1]);
            }
        }
        return dp[n][m];
        
    }
};